### ISSUE TITLE
GPU memory during inference

### ISSUE THREAD

**Issue by jrdimale** [xuelunshen/gim#76]
Thanks for your work!

I would like to inquire about the possibility of optimising the inference speed and GPU memory usage of the model.

I saw in previous issues that you are considering it, but have seen no updates.

Is it possible? What should be the way to proceed to reduce it (a lot)? 

cheers

**Comment by xuelunshen** [comment]
hello, @jrdimale , This matter has not yet made progress. However, I can give you a tip to save GPU memory, such as reducing the size of the input image can greatly reduce the consumption of GPU memory.

**Comment by jrdimale** [comment]
Hi! Indeed I tried reducing the input size of the image, but the GPU consumed during inference is almost the same.
I tried with 480x640, as well as 672x896.
Im using GIMdkm, as its the one that provided the best results in my dataset. Is there another approach to leverage to reduce as much as possible the memory used?

**Comment by xuelunshen** [comment]
Could it be this piece of code causing the issue? `demo.py` changes the size of the input image. You can comment out this section.
https://github.com/xuelunshen/gim/blob/f09105a0555eef5c93db032d97b9e3d8a4cccb20/demo.py#L425-L428

Other methods you could try include:
1. Save the result of `model.match` to a file, then clear the GPU memory before proceeding with `model.sample`.
2. Reduce the input parameter 5000 for `model.sample`.

https://github.com/xuelunshen/gim/blob/f09105a0555eef5c93db032d97b9e3d8a4cccb20/demo.py#L432-L433

**Comment by jrdimale** [comment]
Hi!

I tried multiple configurations and memory is still very high, above 12GB despite making the images smaller and changing this padding.

Furthermore, I'm not planning to use the code in demo.py, actually I use the model in another script by just loading the model with the weights from gim_dkm_100h.ckpt, and performing inference. 

I instantiate a DKMv3 object with the desired h and w, then load the weights and run inference. Is there any change I can do to lower the model gpu usage drastically?

Thanks for your time

**Comment by xuelunshen** [comment]
I see. I think there isn't a very convenient method to quickly achieve your needs right now. The more troublesome approach I can think of is to quantize the model or convert it to TensorRT.

**Comment by jrdimale** [comment]
Have you tried quantization in any of the trained models? Or conversion to TensorRT/ONNXRT? 

If not, do you have any suggestions on how to proceed?

**Comment by xuelunshen** [comment]
Sorry, I'm not familiar with this and haven't tried it before. You need to search Internet to learn the relevant information.

**Comment by laclouis5** [comment]
@jrdimale The most memory hungry of GIM-DKM is in this function:

https://github.com/xuelunshen/gim/blob/f09105a0555eef5c93db032d97b9e3d8a4cccb20/networks/dkm/utils/local_correlation.py#L5-L40

It allocates a very large intermediate tensor `window_feature` of shape `(B, C, H, W, (2*r + 1)**2)` where WxH is the input image sizes and r can range from 3 to 7.

This intermediate tensor can be avoided by chunking the operation:

```python
def local_correlation(
    feature0: torch.Tensor,
    feature1: torch.Tensor,
    local_radius: int,
    padding_mode: str = "zeros",
    flow: torch.Tensor | None = None,
) -> torch.Tensor:
    device = feature0.device
    b, c, h, w = feature0.shape
    r = local_radius
    K = (2 * r + 1) ** 2

    if flow is None:
        ys = torch.linspace(
            -1 + 1 / h, 1 - 1 / h, h, device=device, dtype=feature0.dtype
        )
        xs = torch.linspace(
            -1 + 1 / w, 1 - 1 / w, w, device=device, dtype=feature0.dtype
        )
        gy, gx = torch.meshgrid(ys, xs, indexing="ij")
        base = torch.stack((gx, gy), dim=-1)[None].expand(b, h, w, 2)
    else:
        base = flow.permute(0, 2, 3, 1)

    dx = torch.linspace(
        -2 * r / w, 2 * r / w, 2 * r + 1, device=device, dtype=feature0.dtype
    )
    dy = torch.linspace(
        -2 * r / h, 2 * r / h, 2 * r + 1, device=device, dtype=feature0.dtype
    )
    off_y, off_x = torch.meshgrid(dy, dx, indexing="ij")
    offsets = torch.stack((off_x.reshape(-1), off_y.reshape(-1)), dim=-1)

    out = torch.empty((b, K, h, w), device=device, dtype=feature0.dtype)

    for k in range(K):
        grid = base + offsets[k].view(1, 1, 1, 2)

        samp = F.grid_sample(
            feature1, grid, padding_mode=padding_mode, align_corners=False
        )

        out[:, k] = (feature0 * samp).sum(dim=1)

    return out / (c**0.5)
```

This implementation is not that slower (2 second on the demo images on a NVIDIA RTX 3070 GPU) and it fits on 8GB or VRAM. You can even reduce the VRAM usage down to 6GB using FP16 mixed-precision on some memory heavy modules. Those modifications do not seem to alter the accuracy.

**Comment by xuelunshen** [comment]
@laclouis5 wow, cool, bro
